Fix Hook Test Names

juergba opened the pull request:
### Description of bug
Mocha shows an incorrect test-title if a nested `before` hook fails. For `after` hooks the second part of the message is completely missing.
```js
describe('Top-level describe', () => {
  it('Top-level it', () => { });       // the failing before-hook is pointing to this test
  describe('Nested describe', () => {
    before(() => {
      throw new Error();
    });
    it('Nested it', () => { });        // but it should point to this test
  });
});
```
```
Top-level describe
    ✓ Top-level it
    Nested describe
      1) "before all" hook for "Top-level it"       <<<incorrect
```
Root problem: see [#1638](https://github.com/mochajs/mocha/issues/1638). `hook.ctx.currentTest` is set incorrectly in the hook before emitting the hook event:
- `before` hook: currentTest is undefined or points to the outer test already past successfully.
The runner hasn't reached the next test to come yet, so it's too early to derive currentTest from the runner. Instead the first test of hook's parent is used.
- `after` hook: currentTest is undefined.
The runner has already completed the latest test and runner.currentTest is deleted. Instead the last test of hook's parent is used.

### Description of the Change
#### 1. Setting currentTest correctly
- `before` hook: hook.ctx.currentTest is set to the first test of the parent suite
- `after hook`: hook.ctx.currentTest is set to the last test of the parent suite
- for nested suites just the current suite is searched for tests to come. When no tests can be found, currentTest remains undefined.

#### 2. Creating title in hook failure message
A correct `hook.ctx.currentTest` renders a correct title. 
When no currentTest can be found, the title of the parent suite is used for the hook failure message.

### Alternate Designs
PR [#3333](https://github.com/mochajs/mocha/pull/3333)

### Benefits

This is a rather old bug back to 2015.

### Applicable issues
closes [#1638](https://github.com/mochajs/mocha/issues/1638)
closes [#3291](https://github.com/mochajs/mocha/issues/3291)
closes [#2134](https://github.com/mochajs/mocha/issues/2134)

**EDITED by @boneskull**: added js syntax highlighting, comments

boneskull commented on lib/runner.js:
```diff
@@ -273,6 +273,10 @@ Runner.prototype.failHook = function(hook, err) {
     hook.title =
       hook.originalTitle + ' for "' + hook.ctx.currentTest.title + '"';
+  } else {
+    var parent = hook.parent ? hook.parent.title : 'root';
```
this is incorrect; your original code in 5ea6e653 was fine.  The test failures in `test/unit/runner.spec.js` are due to bad test fixtures--`parent` is a required property of any `Runnable`.

You can fix the tests by ensuring every newly-instantiated `Test` belongs to a `Suite`, like this:

```js
it('should increment .failures', function() {
  expect(runner.failures, 'to be', 0);
  var test1 = new Test('fail hook 1', noop);
  var test2 = new Test('fail hook 2', noop);
  suite.addTest(test1);
  suite.addTest(test2);
  runner.failHook(test1, {});
  expect(runner.failures, 'to be', 1);
  runner.failHook(test2, {});
  expect(runner.failures, 'to be', 2);
});
```

boneskull replied:
that said, you will need to address the fact that the root suite a title of an empty string. :smile:

plroebuck replied:
But suite `title` having an empty string value isn't set in stone, just the default. The root suite _should_ be identified by presence of truthy `root` property.

boneskull replied:
my comment was in reference to how the error message is getting concatenated, anyway.  an empty  `title` would need to be taken into consideration.  

the root suite having an empty string title sounds like one of those things that other code makes assumptions about.

boneskull commented on test/integration/fixtures/current-test-title.fixture.js:
```diff
@@ -0,0 +1,47 @@
+'use strict';
+
+function getTitle() {
+  return this.currentTest ? this.currentTest.title : undefined;
```
maybe more straightforward to skip `Function.prototype.call` altogether:

```js
function getTitle(hook) {
  return hook.currentTest && hook.currentTest.title;
}

before(function() {
  getTitle(this);
});
```

when coerced to a string you're looking at `'undefined'` in your implementation or `false` in mine; either way it should cause a test failure. you'll need to update your test using this of course

boneskull commented on test/integration/hook-err.spec.js:
```diff
@@ -18,7 +18,34 @@ describe('hook error handling', function() {
     before(run('hooks/before-hook-error-tip.fixture.js', onlyErrorTitle()));
     it('should verify results', function() {
-      assert.deepEqual(lines, ['1) spec 2', '"before all" hook:']);
+      assert.deepEqual(lines, [
```
I'm not sure why (can't remember?) we're using `assert` in these tests, but please use `expect` like in the other tests.  if that fails for some reason, then we'll know why :smile:

You don't need to convert all tests in these files to use `expect` (unless you want to)

juergba replied:
@boneskull I'm trying to change "hook-err.spec.js" and to use `expect` as shown in f.e. "options.spec.js". As I understood so far, the key difference is: the "options.\*.fictures.js" don't use `console.log()` while "hooks.\*.fictures.js" do.
In order to use `expect` I need to call (helpers.js)runMochaJSON which results in a parsing error when using `console.log`. Probably I can change most of the tests accordingly, but there will be some depending on the `console.log` output.
Since JSON does not accept comments `//` I don't see a possibility to get around.
Do you have any tips?

boneskull replied:
I think you figured this out?

boneskull commented on test/integration/multiple-done.spec.js:
```diff
@@ -84,7 +84,10 @@ describe('multiple calls to done()', function() {
 
     it('correctly attributes the error', function() {
-      assert.equal(res.failures[0].fullTitle, 'suite "before all" hook');
+      assert.equal(
```
pls try to use `expect`

juergba commented on lib/runner.js:
```diff
@@ -272,7 +272,11 @@ Runner.prototype.failHook = function(hook, err) {
+      ? hook.parent.title
+      : hook.parent.root && !hook.parent.parent
+        ? 'root'
+        : '';
```
`hook.parent.root` is not reliable since any suite with an empty string title "" is marked as root. That's why I'm testing for `hook.parent.parent`, also

plroebuck replied:
Sounds like YA bug...

plroebuck replied:
> `hook.parent.root` is not reliable since any suite with an empty string title "" is marked as root.

Sounds like YA bug...

juergba replied:
see PR #3632

plroebuck replied:
Should the root suite be output as maybe `'{root}'` instead to designate its implied title (rather than an explicit one).

juergba commented:
@boneskull Could you have another look, please? Thanks.

plroebuck commented on test/integration/fixtures/hooks/after-hook-async-error.fixture.js:
```diff
@@ -2,20 +2,13 @@
-  it('should be called because error is in after hook', function () {
-    console.log('test 2');
-  });
+  it('test 1 - should be called', function () {});
```
why was all the context from the test descriptions removed?

plroebuck commented:
What's the status on this PR now that the "just one root" PR is done?

juergba commented:
@plroebuck I will rebase it and revert all changes to test files not directly linked to this PR ("pls try to use expect").

plroebuck commented on test/integration/fixtures/hooks/before-hook-deepnested-error.fixture.js:
```diff
@@ -0,0 +1,13 @@
+      throw new Error('before hook nested error');
+    });
+    describe('spec 3 nested', function () { 
+        it('it nested - this title should not be used', function () { });
```
dedent 2 spaces

plroebuck commented on test/integration/fixtures/hooks/multiple-hook-async-error.fixture.js:
```diff
@@ -1,141 +1,160 @@
-  console.log('root after each');
+after('root after', function () {
+  runOrder.push('root after');
+  assert.deepEqual(runOrder, [
```
Do your assertion check in a separate `after`.

plroebuck replied:
Where did this go?

juergba replied:
I reversed my changes to this file, out of topic.

plroebuck commented on test/integration/fixtures/hooks/multiple-hook-error.fixture.js:
```diff
@@ -1,131 +1,150 @@
-  console.log('root after each');
+after('root after', function () {
+  runOrder.push('root after');
+  assert.deepEqual(runOrder, [
```
same as previous comment

plroebuck commented on test/unit/runner.spec.js:
```diff
@@ -344,9 +344,13 @@ describe('Runner', function() {
+      var test2 = new Test('fail hook 2', noop);
+      suite.addTest(test1);
+      suite.addTest(test2);
+      runner.failHook(test1, {});
```
Shouldn't second argument be an `Error`?

juergba replied:
It's an existing test, I did not change that part. It works fine like this.

plroebuck replied:
Does this work?
```js
runner.failHook(test1, new Error('bummer'));
```

plroebuck replied:
Same as comment below -- create `Error`.

boneskull replied:
I was going to say it wasn't worth requesting, but @plroebuck changed my mind.

It *is* irrelevant to the test, but the test becomes more tightly coupled to the internal implementation if we don't pass an `Error`.  probably best to change it.

juergba commented:
@plroebuck thanks for your feedback. 
@boneskull 
I rebased this branch, adapted some hook titles in existing tests and added my new tests. I reverted many changes to existing tests in order to keep this PR lean.
Could you review, please?

plroebuck commented on lib/runner.js:
```diff
@@ -265,10 +265,17 @@ Runner.prototype.fail = function(test, err) {
+      : hook.parent.root
+      ? '{root}'
+      : '';
+    hook.title = hook.originalTitle + ' in "' + parent + '"';
```
Venturing into ternary op abuse...
```js
    var parent = hook.parent.title
      ? hook.parent.title
      : hook.parent.root
        ? '{root}'
        : '';
```

I indented the latter -- if prettier did this, it's borken...

juergba replied:
I had to use `// prettier-ignore`

plroebuck replied:
Chris won't go for that (turning off prettier).
By reformulate, I meant use two variables or `if` logic -- combining two (or more) ternary operators on a single variable is too much added complexity (metrics-wise) for a single variable (IMO).

boneskull replied:
guilty as charged. 

the idea with prettier is to just let it do whatever it's going to do.  sometimes it might format something a way you don't like, or I don't like, or maybe *nobody* likes, but the intent is to make formatting a non-issue by complete, final, *mindless* adherence to whatever it does.  when we pray at the unholy altar of prettier, it demands no less than total submission.

We *do* have agency when using nested ternaries though.  I try to avoid them in my own code.  but if I've requested people eliminate them in the past, I think I'm over it now.  there are more important things to worry about.

plroebuck commented on test/unit/runner.spec.js:
```diff
@@ -375,6 +380,7 @@ describe('Runner', function() {
     it('should not emit "end" if suite bail is not true', function(done) {
       var hook = new Hook();
+      hook.parent = suite;
       var err = {};
```
Why is this faking `Error`s instead of making them?

juergba replied:
It works both ways.

plroebuck replied:
Then for clarity of intent, create an `Error` instance rather than relying on internal processing to do it for you...

plroebuck commented on lib/runner.js:
```diff
@@ -265,10 +265,17 @@ Runner.prototype.fail = function(test, err) {
     hook.title =
       hook.originalTitle + ' for "' + hook.ctx.currentTest.title + '"';
+  } else {
+    var parent = hook.parent.title
```
`parentTitle`?

juergba replied:
`hook.parent.title` is correct, it's the suite title.

plroebuck replied:
Was in reference to `var parent`

plroebuck commented on test/integration/hook-err.spec.js:
```diff
@@ -18,7 +19,72 @@ describe('hook error handling', function() {
+  describe('before hook root error', function() {
+    it('should verify results', function(done) {
+      runJSON('hooks/before-hook-root-error.fixture.js', [], function(
+        err,
```
Suggest minor change to avoid ^%*^&% prettier reformatting...
```js
      var fixture = 'hooks/before-hook-root-error.fixture.js';
      runJSON(fixture, [], function(err, res) {
```

juergba replied:
I inserted `// eslint-disable-next-line`

plroebuck replied:
Please remove the `eslint` directive -- we use them for enforcement, not formatting

boneskull commented on lib/runner.js:
```diff
@@ -294,7 +302,13 @@ Runner.prototype.hook = function(name, fn) {
     self.currentRunnable = hook;
 
-    hook.ctx.currentTest = self.test;
+    if (name === 'beforeAll') {
```
note to self: replace with constant, see #3655

boneskull commented on test/integration/hook-err.spec.js:
```diff
@@ -36,6 +96,55 @@ describe('hook error handling', function() {
+
+  describe('after hook deepnested error', function() {
+    it('should verify results', function(done) {
+      // eslint-disable-next-line
```
what's avoided here?

plroebuck commented on lib/runner.js:
```diff
@@ -265,10 +265,17 @@ Runner.prototype.fail = function(test, err) {
+      ? hook.parent.title
+      : hook.parent.root
+      ? '{root}'
+      : '';
```
I asked before... please reformulate this. Since prettier denies us the ability to indent the nested ternary, we should not use them. Nesting them is considered a code smell in many languages. [This article](https://medium.com/javascript-scene/nested-ternaries-are-great-361bddd0f340) says they're great, but almost _all_ the commenters disagree.

```js
    var parentTitle;
    if (hook.parent.title) {
      parentTitle = hook.parent.title;
    } else {
      parentTitle = hook.parent.root
        ? '{root}'
        : '';
    }
```

juergba replied:
@plroebuck I usually expect more tolerance against other's coding style, but I will make this change for you. This PR was my first one and you have supported me from beginning. So thanks for that!
Tomorrow I'm going to merge this PR.

boneskull replied:
agree. @mochajs/core 	please dial it back on code style comments.  our linters and formatters are more than up to the task.  if they fall short, we can create new rules to extend functionality.

plroebuck commented on test/integration/fixtures/options/bail-with-after.fixture.js:
```diff
@@ -50,7 +50,7 @@ describe('suite2', function () {
   beforeEach('beforeEach suite2', function () {});
   it('test suite2', function () {
-    runOrder.push('test suite2 - should not run');
+    console.log('test suite2 - should not run');
```
Shouldn't this be an assertion?

juergba replied:
`describe('suite2', function () {...} )` is skipped by the `--bail` flag. So this blog does never run.

plroebuck commented on test/integration/fixtures/options/bail-with-before.fixture.js:
```diff
@@ -37,7 +37,7 @@ describe('suite2', function () {
   beforeEach('beforeEach suite2', function () {});
   it('test suite2', function () {
-    runOrder.push('test suite2 - should not run');
+    console.log('test suite2 - should not run');
```
Shouldn't this be an assertion?

juergba replied:
describe('suite2', function () {...} ) is skipped by the --bail flag. So this blog does never run.

plroebuck commented:
Could you check this against the test cases from #3333?
Believe your PR should cover Heather's tail patch, but would like to be sure.

juergba commented on test/integration/fixtures/hooks/before-hook-root-error.fixture.js:
```diff
@@ -0,0 +1,9 @@
+});
+
+describe('spec 1', function () {
+  it('should not be called', function () { });
```
The assertions are done in "hook-err.spec.js":
```
expect(res, 'to have failed with error', 'before hook root error')
          .and('to have failed test', '"before all" hook in "{root}"')
          .and('to have passed test count', 0);
```

juergba commented:
I checked the tests of #3333, they are covered with this PR and their output is correct.
